http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/listener/WebSocketSourceHandler.java

```diff
@@ -0,0 +1,183 @@
+            String reasonText = closeWebSocketFrame.reasonText();
+            int statusCode = closeWebSocketFrame.statusCode();
+            ctx.channel().close();
+            cMsg = new StatusCarbonMessage(org.wso2.carbon.messaging.Constants.STATUS_CLOSE, statusCode, reasonText);
+
+        } else if (msg instanceof PongWebSocketFrame) {
+            //Control message for WebSocket is Pong Message
+            PongWebSocketFrame pongWebSocketFrame = (PongWebSocketFrame) msg;
+            boolean finalFragment = pongWebSocketFrame.isFinalFragment();
+            ByteBuf byteBuf = pongWebSocketFrame.content();
+            ByteBuffer byteBuffer = byteBuf.nioBuffer();
+            cMsg = new ControlCarbonMessage(byteBuffer, finalFragment);
+            byteBuf.release();
+
+        }
+
+        setupCarbonMessage(ctx);
+        publishToMessageProcessor(cMsg);
+    }
+
+
+    /*
+    Carbon Message is published to registered message processor
+    Message Processor should return transport thread immediately
+     */
+    @Override
+    protected void publishToMessageProcessor(CarbonMessage cMsg) {
+        if (HTTPTransportContextHolder.getInstance().getHandlerExecutor() != null) {
+            HTTPTransportContextHolder.getInstance().getHandlerExecutor().executeAtSourceRequestReceiving(cMsg);
+        }
+
+        CarbonMessageProcessor carbonMessageProcessor = HTTPTransportContextHolder.getInstance()
+                .getMessageProcessor();
+        if (carbonMessageProcessor != null) {
+            try {
+                carbonMessageProcessor.receive(cMsg, new ResponseCallback(this.ctx));
+            } catch (Exception e) {
+                LOGGER.error("Error while submitting CarbonMessage to CarbonMessageProcessor.", e);
+            }
+        } else {
+            LOGGER.error("Cannot find registered MessageProcessor to forward the message.");
+        }
+
+    }
+
+    /*
+    This method runs when initiating a connection.
+     */
+    private void sendOnOpenMessage(ChannelHandlerContext ctx, boolean isSecured, String uri) {
+        cMsg = new StatusCarbonMessage(org.wso2.carbon.messaging.Constants.STATUS_OPEN, 0, null);
+        setupCarbonMessage(ctx);
+        Session session = new WebSocketSessionImpl(ctx, isSecured, uri);
+        cMsg.setProperty(Constants.WEBSOCKET_SESSION, session);
+        cMsg.setProperty(Constants.CHANNEL_ID, channelId);
+        cMsg.setProperty(Constants.CONNECTION, Constants.UPGRADE);
+        cMsg.setProperty(Constants.UPGRADE, Constants.WEBSOCKET_UPGRADE);
+        publishToMessageProcessor(cMsg);
+    }
+
+    /*
+     Extract all the necessary properties from ChannelHandlerContext
+     Add them into a CarbonMessage
+     Note : This method only add details of ChannelHandlerContext to the CarbonMessage
+     Adding other custom details should be done separately
+     @return basic CarbonMessage with necessary details
+     */
+    private void setupCarbonMessage(ChannelHandlerContext ctx) {
```

Comment:
Since it is a private method I didn't put a java doc. Should we put a java doc for private methods too?